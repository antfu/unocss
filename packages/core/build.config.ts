import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  entries: [
    'src/index',
  ],
  clean: true,
  declaration: true,
  externals: [
    'unconfig',
    'magic-string-extra',
  ],
  rollup: {
    emitCJS: true,
  },
})
