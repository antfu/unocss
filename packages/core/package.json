{
  "name": "@unocss/core",
  "version": "0.26.4",
  "description": "The instant on-demand Atomic CSS engine.",
  "keywords": [
    "unocss",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "homepage": "https://github.com/unocss/unocss/tree/main/packages/core#readme",
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss.git",
    "directory": "packages/core"
  },
  "funding": "https://github.com/sponsors/antfu",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    }
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub"
  },
  "devDependencies": {
    "magic-string": "^0.25.7",
    "unconfig": "^0.3.1"
  }
}
