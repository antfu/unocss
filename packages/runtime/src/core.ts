import { initUnocssRuntime } from './factory'

initUnocssRuntime()
