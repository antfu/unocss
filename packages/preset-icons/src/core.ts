import type { DynamicMatcher, Preset, Rule } from '@unocss/core'
import { warnOnce } from '@unocss/core'
import type {
  IconifyLoaderOptions,
  UniversalIconLoader,
} from '@iconify/utils/lib/loader/types'
import { encodeSvgForCss } from '@iconify/utils/lib/svg/encode-svg-for-css'

import { mergeIconProps, trimSVG } from '@iconify/utils'
import type { CSSSVGSprites, IconsOptions } from './types'
import { createAsyncSpriteIconsFactory, createUint8ArraySprite } from './create-sprite'

const COLLECTION_NAME_PARTS_MAX = 3

export { IconsOptions }

/* TODO BEGIN-CLEANUP: types from @iconify/utils: remove them once published */
interface GeneratedSpriteData {
  name: string
  asset: Uint8Array
}
/* TODO END-CLEANUP: types from @iconify/utils: remove them once published */

interface PresetIcons extends Preset {
  generateCSSSVGSprites: () => AsyncIterableIterator<GeneratedSpriteData>
  createCSSSVGSprite: (collection: string) => Promise<Uint8Array | undefined>
}

export function createPresetIcons(lookupIconLoader: (options: IconsOptions) => Promise<UniversalIconLoader>) {
  return function presetIcons(options: IconsOptions = {}): Preset {
    const {
      scale = 1,
      mode = 'auto',
      prefix = 'i-',
      warn = false,
      collections: customCollections,
      extraProperties = {},
      customizations = {},
      autoInstall = false,
      layer = 'icons',
      unit,
      sprites,
    } = options

    const loaderOptions: IconifyLoaderOptions = {
      addXmlNs: true,
      scale,
      customCollections,
      autoInstall,
      // avoid warn from @iconify/loader: we'll warn below if not found
      warn: undefined,
      customizations: {
        ...customizations,
        additionalProps: { ...extraProperties },
        trimCustomSvg: true,
        async iconCustomizer(collection, icon, props) {
          await customizations.iconCustomizer?.(collection, icon, props)
          if (unit) {
            if (!props.width)
              props.width = `${scale}${unit}`
            if (!props.height)
              props.height = `${scale}${unit}`
          }
        },
      },
    }

    const rules: Rule[] = [[
      /^([a-z0-9:_-]+)(?:\?(mask|bg|auto))?$/,
      createDynamicMatcher(warn, mode, loaderOptions, iconLoaderResolver),
      { layer, prefix },
    ]]

    if (sprites) {
      rules.push([
        /^([a-z0-9:_-]+)(?:\?(mask|bg|auto))?$/,
        createDynamicMatcher(warn, sprites.mode ?? mode, loaderOptions, iconLoaderResolver, sprites),
        { layer, prefix: sprites.prefix ?? 'sprite-' },
      ])
    }

    let iconLoader: UniversalIconLoader

    async function iconLoaderResolver() {
      iconLoader = iconLoader || await lookupIconLoader(options)
      return iconLoader
    }

    return <PresetIcons>{
      name: '@unocss/preset-icons',
      enforce: 'pre',
      options,
      layers: { icons: -30 },
      rules,
      generateCSSSVGSprites: createGenerateCSSSVGSprites(sprites),
      createCSSSVGSprite: createCSSSVGSpriteLoader(sprites),
    }
  }
}

export function combineLoaders(loaders: UniversalIconLoader[]) {
  return (async (...args) => {
    for (const loader of loaders) {
      if (!loader)
        continue
      const result = await loader(...args)
      if (result)
        return result
    }
  }) as UniversalIconLoader
}

function createDynamicMatcher(
  warn: boolean,
  mode: string,
  loaderOptions: IconifyLoaderOptions,
  iconLoaderResolver: () => Promise<UniversalIconLoader>,
  sprites?: CSSSVGSprites,
) {
  return <DynamicMatcher>(async ([full, body, _mode = mode]) => {
    let collection = ''
    let name = ''
    let svg: string | undefined

    const iconLoader = await iconLoaderResolver()

    const usedProps = {}
    if (body.includes(':')) {
      [collection, name] = body.split(':')
      svg = sprites
        ? await loadSvgFromSprite(collection, name, sprites, { ...loaderOptions, usedProps })
        : await iconLoader(collection, name, { ...loaderOptions, usedProps })
    }
    else {
      const parts = body.split(/-/g)
      for (let i = COLLECTION_NAME_PARTS_MAX; i >= 1; i--) {
        collection = parts.slice(0, i).join('-')
        name = parts.slice(i).join('-')
        svg = sprites
          ? await loadSvgFromSprite(collection, name, sprites, { ...loaderOptions, usedProps })
          : await iconLoader(collection, name, { ...loaderOptions, usedProps })
        if (svg)
          break
      }
    }

    if (!svg) {
      if (warn)
        warnOnce(`failed to load icon "${full}"`)
      return
    }

    let url: string
    if (sprites)
      url = `url("unocss-${collection}-sprite.svg#${name}")`
    else
      url = `url("data:image/svg+xml;utf8,${encodeSvgForCss(svg)}")`

    if (_mode === 'auto')
      _mode = svg.includes('currentColor') ? 'mask' : 'bg'

    if (_mode === 'mask') {
      // Thanks to https://codepen.io/noahblon/post/coloring-svgs-in-css-background-images
      return {
        '--un-icon': url,
        '-webkit-mask': 'var(--un-icon) no-repeat',
        'mask': 'var(--un-icon) no-repeat',
        '-webkit-mask-size': '100% 100%',
        'mask-size': '100% 100%',
        'background-color': 'currentColor',
        // for Safari https://github.com/elk-zone/elk/pull/264
        'color': 'inherit',
        ...usedProps,
      }
    }
    else {
      return {
        'background': `${url} no-repeat`,
        'background-size': '100% 100%',
        'background-color': 'transparent',
        ...usedProps,
      }
    }
  })
}

/* TODO BEGIN-CLEANUP: types from @iconify/utils: remove them once published */
function createCSSSVGSpriteLoader(sprites?: CSSSVGSprites) {
  return async (collection: string) => {
    const collections = sprites?.sprites[collection]
    if (!collections)
      return undefined

    return await createUint8ArraySprite(
      collection,
      createAsyncSpriteIconsFactory(collections),
      sprites?.warn ?? false,
    )
  }
}

function createGenerateCSSSVGSprites(sprites?: CSSSVGSprites) {
  return async function* () {
    if (sprites) {
      const warn = sprites.warn ?? false
      for (const [collection, collections] of Object.entries(sprites.sprites)) {
        yield <GeneratedSpriteData>{
          name: collection,
          asset: await createUint8ArraySprite(
            collection,
            createAsyncSpriteIconsFactory(collections),
            warn,
          ),
        }
      }
    }
  }
}

async function customizeSpriteIcon(
  collection: string,
  icon: string,
  options: IconifyLoaderOptions,
  svg?: string,
) {
  if (!svg)
    return svg

  const cleanupIdx = svg.indexOf('<svg')
  if (cleanupIdx > 0)
    svg = svg.slice(cleanupIdx)

  const { transform } = options?.customizations ?? {}
  svg
      = typeof transform === 'function'
      ? await transform(svg, collection, icon)
      : svg

  if (!svg.startsWith('<svg')) {
    console.warn(
        `Custom icon "${icon}" in "${collection}" is not a valid SVG`,
    )
    return svg
  }

  return await mergeIconProps(
    options?.customizations?.trimCustomSvg === true
      ? trimSVG(svg)
      : svg,
    collection,
    icon,
    options,
    undefined,
  )
}

async function loadSvgFromSprite(
  collectionName: string,
  icon: string,
  sprites: CSSSVGSprites,
  options: IconifyLoaderOptions,
) {
  const collections = sprites.sprites[collectionName]
  if (!collections)
    return

  const collectionsArray = Array.isArray(collections)
    ? collections
    : [collections]

  for (const collection of collectionsArray) {
    if (Array.isArray(collection)) {
      for (const spriteIcon of collection) {
        if (spriteIcon.name === icon)
          return await customizeSpriteIcon(collectionName, icon, options, spriteIcon.svg)
      }
    }
    else if ('svg' in collection) {
      if (collection.name === icon)
        return await customizeSpriteIcon(collectionName, icon, options, collection.svg)
    }
    else if (typeof collection === 'function') {
      for await (const spriteIcon of collection()) {
        if (spriteIcon.name === icon)
          return await customizeSpriteIcon(collectionName, icon, options, spriteIcon.svg)
      }
    }
    else {
      for await (const spriteIcon of collection[Symbol.asyncIterator]()) {
        if (spriteIcon.name === icon)
          return await customizeSpriteIcon(collectionName, icon, options, spriteIcon.svg)
      }
    }
  }
}
/* TODO END-CLEANUP: types from @iconify/utils: remove them once published */
