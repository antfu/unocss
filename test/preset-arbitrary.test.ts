import { createGenerator, escapeSelector } from '@unocss/core'
import presetWind from '@unocss/preset-wind'
import { describe, expect, test } from 'vitest'
import presetArbitrary from '@unocss/preset-arbitrary'
import { presetArbitraryTargets } from './assets/preset-arbitrary-target'

const uno = createGenerator({
  presets: [
    presetWind(),
    presetArbitrary(),
  ],
})

describe('preset-arbitrary', () => {
  test('targets', async () => {
    const code = presetArbitraryTargets.join(' ')
    const { css } = await uno.generate(code)
    const { css: css2 } = await uno.generate(code)

    expect(uno.config.extractors).toHaveLength(1)
    expect(uno.config.extractors[0].name).toBe('split-arbitrary')

    const unmatched = []
    for (const i of presetArbitraryTargets) {
      if (!css.includes(escapeSelector(i)))
        unmatched.push(i)
    }
    expect(unmatched).toEqual([])
    expect(css).toMatchSnapshot()
    expect(css).toEqual(css2)
  })
})
