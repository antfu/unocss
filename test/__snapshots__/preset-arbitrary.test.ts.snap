// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset-arbitrary > targets 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: default */
.\\\\[--css-variable\\\\:\\\\\\"wght\\\\\\"_400\\\\,_\\\\\\"opsz\\\\\\"_14\\\\]{--css-variable:\\"wght\\" 400, \\"opsz\\" 14;}
.\\\\[--escaped\\\\\\\\\\\\~variable\\\\\\\\\\\\:\\\\:100\\\\%\\\\]{--escaped\\\\~variable\\\\::100%;}
.\\\\[a\\\\:b\\\\]{a:b;}
.\\\\[background-image\\\\:url\\\\(star_transparent\\\\.gif\\\\)\\\\,_url\\\\(cat_front\\\\.png\\\\)\\\\]{background-image:url(star_transparent.gif), url(cat_front.png);}
.\\\\[content\\\\:attr\\\\(attr_content\\\\)\\\\]{content:attr(attr content);}
.\\\\[content\\\\:attr\\\\(attr\\\\\\\\_content\\\\)\\\\]{content:attr(attr_content);}
.\\\\[font-family\\\\:\\\\'Inter\\\\'\\\\,_sans-serif\\\\]{font-family:'Inter', sans-serif;}
.\\\\[font-family\\\\:var\\\\(--font-family\\\\)\\\\]{font-family:var(--font-family);}
.\\\\[font-feature-settings\\\\:\\\\'cv02\\\\'\\\\,\\\\'cv03\\\\'\\\\,\\\\'cv04\\\\'\\\\,\\\\'cv11\\\\'\\\\]{font-feature-settings:'cv02','cv03','cv04','cv11';}
.\\\\[font-variation-settings\\\\:\\\\\\"wght\\\\\\"_400\\\\,_\\\\\\"opsz\\\\\\"_14\\\\]{font-variation-settings:\\"wght\\" 400, \\"opsz\\" 14;}
.\\\\[margin\\\\:logical_1rem_2rem_3rem\\\\]{margin:logical 1rem 2rem 3rem;}
.\\\\[\\\\&_\\\\&\\\\]\\\\:m-13 .\\\\[\\\\&_\\\\&\\\\]\\\\:m-13{margin:3.25rem;}
.\\\\[\\\\&\\\\:nth-child\\\\(2\\\\)\\\\]\\\\:m-10:nth-child(2){margin:2.5rem;}
.\\\\[\\\\&\\\\[open\\\\]\\\\:readonly\\\\]\\\\:m-16[open]:readonly{margin:4rem;}
.\\\\[\\\\&\\\\[open\\\\]\\\\]\\\\:m-14[open]{margin:3.5rem;}
.\\\\[\\\\&\\\\[readonly\\\\]\\\\[disabled\\\\]\\\\]\\\\:m-15[readonly][disabled]{margin:3.75rem;}
.\\\\[\\\\&\\\\>\\\\*\\\\]\\\\:m-11>*{margin:2.75rem;}
*[open]:readonly .\\\\[\\\\*\\\\[open\\\\]\\\\:readonly_\\\\&\\\\]\\\\:\\\\[\\\\&\\\\[open\\\\]\\\\:disabled\\\\]\\\\:m-17[open]:disabled{margin:4.25rem;}
*>.\\\\[\\\\*\\\\>\\\\&\\\\]\\\\:m-12{margin:3rem;}
.\\\\[\\\\&\\\\[data-active\\\\=\\\\\\"true\\\\\\"\\\\]\\\\]\\\\:bg-red[data-active=\\"true\\"]{--un-bg-opacity:1;background-color:rgba(248,113,113,var(--un-bg-opacity));}
.before\\\\:\\\\[\\\\&\\\\[data-active\\\\=\\\\'true\\\\'\\\\]\\\\]\\\\:content-\\\\[\\\\'test\\\\'\\\\][data-active='true']::before{content:'test';}
.content-\\\\[\\\\'quoted_with_space\\\\'\\\\]{content:'quoted with space';}
.content-\\\\[\\\\'quoted1\\\\'\\\\]{content:'quoted1';}
.content-\\\\[\\\\\\"quoted2\\\\\\"\\\\]{content:\\"quoted2\\";}
@media (hover: hover) and (pointer: fine){
[open] .\\\\@hover\\\\:\\\\[\\\\[open\\\\]_\\\\&\\\\]\\\\:text-blue:hover{--un-text-opacity:1;color:rgba(96,165,250,var(--un-text-opacity));}
}
@supports(display:grid){
.\\\\[\\\\@supports\\\\(display\\\\:grid\\\\)\\\\]\\\\:bg-red\\\\/33{background-color:rgba(248,113,113,0.33);}
*+.\\\\[\\\\@supports\\\\(display\\\\:grid\\\\)\\\\]\\\\:\\\\[\\\\*\\\\+\\\\&\\\\]\\\\:bg-red\\\\/34{background-color:rgba(248,113,113,0.34);}
}"
`;
